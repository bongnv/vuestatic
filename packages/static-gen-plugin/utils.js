const url = require("url");
const cheerio = require("cheerio");

const relativePathsFromHtml = ({ html, currentPath }) => {
  const $ = cheerio.load(html);

  const linkHrefs = $("a[href]")
    .map(function (i, el) {
      return $(el).attr("href");
    })
    .get();

  const iframeSrcs = $("iframe[src]")
    .map(function (i, el) {
      return $(el).attr("src");
    })
    .get();

  return []
    .concat(linkHrefs)
    .concat(iframeSrcs)
    .map(function (href) {
      if (href.indexOf("//") === 0) {
        return null;
      }

      var parsed = url.parse(href);

      if (parsed.protocol || typeof parsed.path !== "string") {
        return null;
      }

      return parsed.path.indexOf("/") === 0
        ? parsed.path
        : url.resolve(currentPath, parsed.path);
    })
    .filter(function (href) {
      return href != null;
    });
};

exports.relativePathsFromHtml = relativePathsFromHtml;
