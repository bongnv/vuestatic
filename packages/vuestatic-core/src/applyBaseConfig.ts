import path from "path";
import webpack from "webpack";
import Config from "webpack-chain";
import MiniCssExtractPlugin from "mini-css-extract-plugin";
import { VueLoaderPlugin } from "vue-loader";
import { CleanWebpackPlugin } from "clean-webpack-plugin";

import type { NormalizedConfig } from "./Execution";

const applyBaseConfig = (
  { isProd, srcDir }: NormalizedConfig,
  webpackConfig: Config,
): void => {
  webpackConfig.mode(isProd ? "production" : "development");
  webpackConfig.module
    .rule("compile-vue")
    .test(/\.vue$/)
    .use("vue-loader")
    .loader("vue-loader");

  webpackConfig.module
    .rule("compile-css")
    .test(/\.css$/)
    .use("mini-css-extract")
    .loader(MiniCssExtractPlugin.loader)
    .end()
    .use("css-loader")
    .loader("css-loader")
    .options({
      importLoaders: 1,
    })
    .end()
    .use("postcss-loader")
    .loader("postcss-loader")
    .end();

  webpackConfig.resolve.modules.add(srcDir).add("node_modules");

  webpackConfig.plugin("vue-loader").use(new VueLoaderPlugin());

  webpackConfig.plugin("mini-css-extract").use(
    new MiniCssExtractPlugin({
      filename: path.join(
        "_assets",
        isProd ? "[name].[contenthash].css" : "[name].css",
      ),
    }),
  );

  webpackConfig
    .plugin("clean")
    .use(new CleanWebpackPlugin({ cleanStaleWebpackAssets: false }));

  webpackConfig.plugin("define").use(
    new webpack.DefinePlugin({
      "process.env.NODE_ENV": isProd ? '"production"' : '"development"',
    }),
  );
};

export = applyBaseConfig;
