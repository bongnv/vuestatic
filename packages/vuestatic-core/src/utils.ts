import path from "path";
import Config from "webpack-chain";

import type { NormalizedConfig, ExecutionConfig } from "./Execution";

export const defaultExecutionConfig = (
  rawConfig: ExecutionConfig = {},
): NormalizedConfig => {
  const baseDir = rawConfig.baseDir || path.resolve(process.cwd());
  const outputDir = rawConfig.outputDir || path.join(baseDir, "dist");
  const serverPath =
    rawConfig.serverPath || path.join(baseDir, ".vuestatic", "server");
  const srcDir = rawConfig.srcDir || path.join(baseDir, "src");
  const plugins = rawConfig.plugins || [];

  return {
    isProd: false,
    baseDir,
    outputDir,
    publicPath: "/",
    serverPath,
    srcDir,
    clientPlugins: [],
    propsHandlers: [],
    plugins: plugins,
    clientWebpackConfig: new Config(),
    serverWebpackConfig: new Config(),
  };
};
