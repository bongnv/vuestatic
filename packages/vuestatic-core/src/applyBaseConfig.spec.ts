import Config from "webpack-chain";
import MiniCssExtractPlugin from "mini-css-extract-plugin";

import applyBaseConfig from "./applyBaseConfig";
import { defaultExecutionConfig } from "./utils";

jest.mock("mini-css-extract-plugin");

test("applyBaseConfig should configure basic configuration", () => {
  const webpackConfig = new Config();
  const config = defaultExecutionConfig();
  applyBaseConfig(config, webpackConfig);
  const serializedConfig = webpackConfig.toConfig();
  expect(serializedConfig).toBeTruthy();
  expect(serializedConfig.mode).toBe("development");
  expect(serializedConfig.module?.rules).toHaveLength(2);
});

test("applyBaseConfig should match snapshot", () => {
  MiniCssExtractPlugin.loader = "mini-css-extract-plugin-loader";
  const webpackConfig = new Config();
  const config = defaultExecutionConfig({
    baseDir: "/",
  });
  applyBaseConfig(config, webpackConfig);
  const serializedConfig = webpackConfig.toConfig();
  expect(serializedConfig).toMatchSnapshot();
});
