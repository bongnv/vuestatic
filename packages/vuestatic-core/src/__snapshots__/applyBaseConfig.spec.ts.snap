// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`applyBaseConfig should match snapshot 1`] = `
Object {
  "mode": "development",
  "module": Object {
    "rules": Array [
      Object {
        "test": /\\\\\\.vue\\$/,
        "use": Array [
          Object {
            "loader": "vue-loader",
          },
        ],
      },
      Object {
        "test": /\\\\\\.css\\$/,
        "use": Array [
          Object {
            "loader": "/Users/van.bong/projects/vuestatic/node_modules/mini-css-extract-plugin/dist/loader.js",
          },
          Object {
            "loader": "css-loader",
            "options": Object {
              "importLoaders": 1,
            },
          },
          Object {
            "loader": "postcss-loader",
          },
        ],
      },
    ],
  },
  "node": Object {
    "setImmediate": false,
  },
  "plugins": Array [
    VueLoaderPlugin {},
    MiniCssExtractPlugin {
      "options": Object {
        "chunkFilename": "_assets/[name].css",
        "filename": "_assets/[name].css",
        "ignoreOrder": false,
        "moduleFilename": [Function],
      },
    },
    CleanWebpackPlugin {
      "apply": [Function],
      "cleanAfterEveryBuildPatterns": Array [],
      "cleanOnceBeforeBuildPatterns": Array [
        "**/*",
      ],
      "cleanStaleWebpackAssets": false,
      "currentAssets": Array [],
      "dangerouslyAllowCleanPatternsOutsideProject": false,
      "dry": false,
      "handleDone": [Function],
      "handleInitial": [Function],
      "initialClean": false,
      "outputPath": "",
      "protectWebpackAssets": true,
      "removeFiles": [Function],
      "verbose": false,
    },
    DefinePlugin {
      "definitions": Object {
        "process.env.NODE_ENV": "\\"development\\"",
      },
    },
  ],
  "resolve": Object {
    "alias": Object {
      "@vuestatic/app": "/Users/van.bong/projects/vuestatic/packages/vuestatic-core/src/app.js",
    },
  },
}
`;
